[flang] Use precompiled headers in Frontend, Lower, Parser, Semantics and Evaluate

mrkajetanp opened the pull request:
Precompiling larger headers can save a lot of compile time across various compilation units.

Selected compile time & memory improvements are as follows:

flang/lib/Parser/Fortran-parsers.cpp:
Elapsed (wall clock) time (h:mm:ss or m:ss): 0:47.31 -> 0:41.68
Maximum resident set size (kbytes): 2062140 -> 1745584

flang/lib/Lower/Bridge.cpp:
Elapsed (wall clock) time (h:mm:ss or m:ss): 1:19.16 -> 0:45.86
Maximum resident set size (kbytes): 3849144 -> 2443476

flang/lib/Lower/PFTBuilder.cpp
Elapsed (wall clock) time (h:mm:ss or m:ss): 1:29.24 -> 1:00.99
Maximum resident set size (kbytes): 4218368 -> 2923128

flang/lib/Lower/Allocatable.cpp
Elapsed (wall clock) time (h:mm:ss or m:ss): 0:53.03 -> 0:22.50
Maximum resident set size (kbytes): 3092840 -> 2116908

flang/lib/Semantics/Semantics.cpp
Elapsed (wall clock) time (h:mm:ss or m:ss): 1:18.75 -> 1:00.20
Maximum resident set size (kbytes): 3527744 -> 2545308

While the newly added precompiled headers are as follows:

Parser:
Elapsed (wall clock) time (h:mm:ss or m:ss): 0:09.62
Maximum resident set size (kbytes): 1034608

Lower:
Elapsed (wall clock) time (h:mm:ss or m:ss): 0:41.33
Maximum resident set size (kbytes): 3615240

Semantics:
Elapsed (wall clock) time (h:mm:ss or m:ss): 0:26.69
Maximum resident set size (kbytes): 2403776

mrkajetanp committed:
[flang] Use precompiled headers in Lower, Parser and Semantics

Precompiling larger headers can save a lot of compile time across
various compilation units.

Selected compile time & memory improvements are as follows:

flang/lib/Parser/Fortran-parsers.cpp:
Elapsed (wall clock) time (h:mm:ss or m:ss): 0:47.31 -> 0:41.68
Maximum resident set size (kbytes): 2062140 -> 1745584

flang/lib/Lower/Bridge.cpp:
Elapsed (wall clock) time (h:mm:ss or m:ss): 1:19.16 -> 0:45.86
Maximum resident set size (kbytes): 3849144 -> 2443476

flang/lib/Lower/PFTBuilder.cpp
Elapsed (wall clock) time (h:mm:ss or m:ss): 1:29.24 -> 1:00.99
Maximum resident set size (kbytes): 4218368 -> 2923128

flang/lib/Lower/Allocatable.cpp
Elapsed (wall clock) time (h:mm:ss or m:ss): 0:53.03 -> 0:22.50
Maximum resident set size (kbytes): 3092840 -> 2116908

flang/lib/Semantics/Semantics.cpp
Elapsed (wall clock) time (h:mm:ss or m:ss): 1:18.75 -> 1:00.20
Maximum resident set size (kbytes): 3527744 -> 2545308

While the newly added precompiled headers are as follows:

Parser:
Elapsed (wall clock) time (h:mm:ss or m:ss): 0:09.62
Maximum resident set size (kbytes): 1034608

Lower:
Elapsed (wall clock) time (h:mm:ss or m:ss): 0:41.33
Maximum resident set size (kbytes): 3615240

Semantics:
Elapsed (wall clock) time (h:mm:ss or m:ss): 0:26.69
Maximum resident set size (kbytes): 2403776

Signed-off-by: Kajetan Puchalski <kajetan.puchalski@arm.com>

jeanPerier commented:
Lowering depends on Semantics and the Parser, and Semantics depends on Parser, have you tried using REUSE_FROM https://cmake.org/cmake/help/latest/command/target_precompile_headers.html#reusing-precompile-headers so that lowering can reuse the precompiled headers too?

Just curious, I am not familiar with that cmake feature at all.

mrkajetanp commented:
> Lowering depends on Semantics and the Parser, and Semantics depends on Parser, have you tried using REUSE_FROM 

We shortly discussed this in a subthread under https://github.com/llvm/llvm-project/pull/130600.
In short, REUSE_FROM is an all-or-nothing kind of situation. With the feature it's not possible to reuse a different precompiled header unit and also add extra ones at the same time, you either re-use and existing one or have to build a new one completely from scratch.

Since the main goal is to reduce the memory usage per compilation thread, I didn't want to put too much into the precompiled headers either. If I tried adding the same Semantics headers on top of the Lower ones then the precompiled header for Lower/ would become probably the biggest TU in all of flang.

klausler commented:
Why not Evaluate?  It might benefit the most of all of these directories.

mrkajetanp commented:
> Why not Evaluate? It might benefit the most of all of these directories.

Thanks for the suggestion! No reason not to indeed, I'll add it.

jeanPerier commented:
What is the impact on the size of the build directories? (I am asking because this[ blog post](https://www.qt.io/blog/2019/08/01/precompiled-headers-and-unity-jumbo-builds-in-upcoming-cmake) is pointing out that the build directory size can increase).

mrkajetanp commented:
> What is the impact on the size of the build directories?

From `du -h`:
440M    lib/Lower/CMakeFiles/FortranLower.dir/cmake_pch.hxx.pch
122M    lib/Parser/CMakeFiles/FortranParser.dir/cmake_pch.hxx.pch
278M    lib/Semantics/CMakeFiles/FortranSemantics.dir/cmake_pch.hxx.pch
164M    lib/Evaluate/CMakeFiles/FortranEvaluate.dir/cmake_pch.hxx.pch

But then at the same time I think some space might be saved in the specific compilation units that now no longer need to include them, so it might not be the biggest deal.

mrkajetanp committed:
[flang] Use precompiled headers in Evaluate

flang/lib/Evaluate/check-expression.cpp:
Elapsed (wall clock) time (h:mm:ss or m:ss): 1:11.91 -> 1:02.29
Maximum resident set size (kbytes): 2710788 -> 2414740

Similar improvements for other compilation units under Evaluate.

cmake_pch.hxx.cxx compilation time:
Elapsed (wall clock) time (h:mm:ss or m:ss): 0:13.93
Maximum resident set size (kbytes): 1492744

Signed-off-by: Kajetan Puchalski <kajetan.puchalski@arm.com>

Meinersbur commented:
Consider that his does  not work with ccache: https://github.com/llvm/llvm-project/pull/130600#issuecomment-2725724370

mgorny reviewed (approved):
Thanks for doing this! I can confirm, that (combined with #131397) I can build Flang with `-j6` — and basically it'd go with `-j12` if not for a few libraries.

DavidTruby reviewed (approved):
LGTM thanks!

mrkajetanp committed:
[flang] Use precompiled headers in Frontend

Reland of #130600 which was reverted on account of waiting for required
ccache compatibility fixes.

Signed-off-by: Kajetan Puchalski <kajetan.puchalski@arm.com>

mrkajetanp commented:
I reverted #130600 as it was causing some buildbot issues in the absence of the ccache fixes.
I've now added the changes in that PR here and rebased this so all the PCH can be merged together once the ccache issues are resolved.

mrkajetanp committed:
[CMake] Disable precompiled headers for ccache builds on Windows

Signed-off-by: Kajetan Puchalski <kajetan.puchalski@arm.com>

mrkajetanp commented:
With #136856 merged there is now proper ccache support for this on Linux. I added a line here to disable pch on Windows for the time being, meaning this should now be ready to go.

Meinersbur commented on llvm/CMakeLists.txt:
```diff
@@ -286,6 +286,9 @@ if(LLVM_CCACHE_BUILD)
     else()
+      # Until a way to reliably configure ccache on Windows is found,
+      # disable precompiled headers for Windows + ccache builds
+      set(CMAKE_DISABLE_PRECOMPILE_HEADERS "ON")
```
Consider emitting a message informing the user if their `CMAKE_DISABLE_PRECOMPILE_HEADERS=OFF` setting was overwritten.

mrkajetanp replied:
Good suggestion, thank you! Will do.

Meinersbur reviewed (approved):
LGTM

mrkajetanp committed:
[CMake] Warn when overriding CMAKE_DISABLE_PRECOMPILED_HEADERS

Signed-off-by: Kajetan Puchalski <kajetan.puchalski@arm.com>

mrkajetanp commented:
Okay, let's see how this goes then.
